reader.go

```diff
@@ -70,8 +112,33 @@ func (r *Reader) Read(p []byte) (n int, err error) {
 // Close closes this Reader on the Stream. This must be called when done with the
 // Reader or else the Stream cannot be Removed.
 func (r *Reader) Close() error {
+	r.close = true
+	r.s.b.Unblock(r.id)
+	r.mu.Lock()
+	r.mu.Unlock()
+	return r.closeUnsafe()
+}
+
+// Cancel closes this Reader on the Stream and fails next reads from reader.
+// It immediately cancels all pending reads
+func (r *Reader) Cancel() error {
+	r.cancel = true
+	r.s.b.Unblock(r.id)
```

Comment:
This will unblock all blocking all Readers (thanks to Broadcast), I wish there was a way to signal a specific one to unblock as unblocking all readers just to check if a single one has canceled is a bit of a pain.

Reply:
Yup, I'm await. But there's no easy way to implement canceling on readers other way, no? I think unblocking all readers upon canceling single one fair tradeoff, especially because they are constantly unblocking on every write. I think the performance won't hurt.

Reply:
There isn't a nice way to do it in Go right now, something like https://github.com/golang/go/issues/16620 would be nice, but it's just a proposal. I'm not criticizing your implementation, just lamenting that I wish this didn't have to wake all blocking reads for nothing (would be great if we had something like cond.SignalWaiter(waiterId) or similar).

The main diff. I feel here is that yes all Writes broadcast, but every woken Reader likely has something to do (read the written data). In this case, it's very unlikely that each woken Reader has something to do.

Do you have an immediate use case for this feature? I'm also wondering about the case where the trigger for Reader.Cancel() might happen for all the Readers, causing n Readers to all call broadcast => n^2 wakeups.

Also, what's your use case for Stream.Cancel feature as well, it's always nice to know the problem we're solving.

Reply:
I'm using stream for work queue where client is sending streaming work request and workers can subscribe for this work request (and perform work). Let's say it's video processing and workers are converting video to multiple resolutions.

Stream.Cancel() is used for cancelling request for work, which should cancel all subscribers as well, so they stop performing work they maybe started already doing (e.g. someone removed video immediately after uploading, before workers converted it).

Workers can fail or hang for some reasons (e.g. broken network or deadlock). Server should note it immediately and re-try work on other node, but most importantly it should not expect failed worker to perform any more Read from the stream, so ignoring unfinished read when e.g. waiting for all readers to finish when doing Remove.

If Stream.Cancel() doesn't work, workers perform unnecessary work.

If Reader.Cancel() doesn't work, server will wait forever for broken workers to finish read and call .Done() on WaitGroup what will never happen.

So all to all, I'm thinking just doing .closeUnsafe() and decrementing waitgroup instead of unlocking reader should suffice.

Reply:
Writing this makes me want a Stream.Wait() method that just wait for all readers to finish, without Removing the file.

Reply:
With Stream.Remove() you get the guarantee that no new Readers will be created, for Stream.Wait() are you okay with the race between the return from Stream.Wait() and the possibility of new Readers being started?

This kind of race is normal with sync.WaitGroup, it's usually prevented by insuring you don't call "Add()" after Wait() might have completed. In this case, it would mean you would not call NextReader() after Wait() has been called.

Reply:
Let's forget about Wait() for now, I'd leave it for later as it needs such discussion